### PR TITLE
Maze/AntMaze issues

### LINKED ISSUE #155: [Bug Report] Maze/AntMaze issues
Hi, I believe I have found a couple of issues in the Maze/AntMaze environments. I have resolved both of these issues in [commit 5573d5e](https://github.com/Farama-Foundation/Gymnasium-Robotics/compare/main...alexdavey:Gymnasium-Robotics:antmaze), and I’m happy to submit a PR.

### 1) AntMaze sparse reward always zero

For continuing tasks in AntMazeEnv, the sparse reward is always zero. This is because at each step, `.compute_terminated()` resets the goal when the Ant is sufficiently close, before the reward is calculated.

Here’s a code example to test this:

```python
import numpy as np
import gymnasium as gym

env = gym.make("AntMaze_UMaze-v3", continuing_task=True, reward_type="sparse")

total_reward = 0

# 1000 "episodes" of 100 steps
for _ in range(1000):
    env.reset()

    for _ in range(100):
        action = env.action_space.sample()
        _, rew, _, _, _ = env.step(action)
        total_reward += rew

print(total_reward)

```

Currently this returns exactly zero since no reward is collected, but placing `.compute_reward()` above `.compute_terminated()` gives a non-zero reward. In PointMaze, this issue was fixed in commit ace181e2a0d83be8940103bbb4d7e917afb30119.

### 2) AntMaze can reset into a terminal state

The Ant will sometimes start within the goal radius. This is because there is a `maze_size_scaling` factor missing in the distance check in `MazeEnv.generate_reset_pos()`.

In AntMaze `maze_size_scaling = 4`, so the xy position noise can be up to 1.0 in each direction. An unfortunate combination of goal noise and reset noise can cause the Ant to start within 0.45 of `self.goal`. This issue does not affect PointMaze because there `maze_size_scaling = 1`.

Here’s a code example to test this:
```python
import numpy as np
import gymnasium as gym

env = gym.make("AntMaze_UMaze-v3", continuing_task=True)

for _ in range(1000):
    obs, _ = env.reset()
    dist = np.linalg.norm(obs["achieved_goal"] - obs["desired_goal"])
    assert dist > 0.45
```

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/gymnasium-robotics/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by alexdavey**
# Description

Fixes #155 

* refactored `compute_terminated` so it is now a pure function, added `update_goal`

* added info["success"] key to AntMaze

* fixed sparse reward issue in AntMaze

* fixed reset issue in Maze

* added test for reset issue

* bumped version to AntMaze-v4

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by rodrigodelazcano** [comment]
We should be keeping previous versions of the environment such as in Gymnasium. You can create different file versions for each environment version `antmaze_v3.py` and `antmaze_v4.py`. The updates in `maze.py` don't have any breaking changes so it doesn't require a version bump.

**Review by rodrigodelazcano — CHANGES_REQUESTED** [review]

**Comment by rodrigodelazcano** [comment]
@alexdavey I think the test is failing because the naming `test_reset.py` can't be repeated.

Actually, I would prefer if the test files have the same structure as the Gymnasium repo. Add two files `tests/envs/maze/point_maze.py` and `tests/envs/maze/ant_maze.py` where we can add tests to each environment.

File: gymnasium_robotics/envs/maze/ant_maze.py [hunk 1]
```diff
@@ -192,10 +192,11 @@ class AntMazeEnv(MazeEnv, EzPickle):
     ```python
     import gymnasium as gym
 
-    env = gym.make('AntMaze_UMaze-v3', max_episode_steps=100)
+    env = gym.make('AntMaze_UMaze-v4', max_episode_steps=100)
     ```
 
     ### Version History
+    * v4: refactor of compute_terminated, fixed sparse reward and reset issues
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/ant_maze.py:199]: Be more verbose

File: gymnasium_robotics/envs/maze/ant_maze.py [hunk 2]
```diff
@@ -271,21 +272,27 @@ def reset(self, *, seed: Optional[int] = None, **kwargs):
 
         obs, info = self.ant_env.reset(seed=seed)
         obs_dict = self._get_obs(obs)
+        info["success"] = bool(
+            np.linalg.norm(obs_dict["achieved_goal"] - self.goal) <= 0.45
+        )
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/ant_maze.py:277]: why is this here?

File: gymnasium_robotics/envs/maze/ant_maze.py [hunk 3]
```diff
@@ -271,21 +272,27 @@ def reset(self, *, seed: Optional[int] = None, **kwargs):
...
 
         return obs_dict, info
 
     def step(self, action):
         ant_obs, _, _, _, info = self.ant_env.step(action)
         obs = self._get_obs(ant_obs)
 
+        info["success"] = bool(np.linalg.norm(obs["achieved_goal"] - self.goal) <= 0.45)
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/ant_maze.py:285]: `info` should be after `truncated`

File: gymnasium_robotics/envs/maze/maze.py [hunk 4]
```diff
@@ -363,19 +365,22 @@ def compute_terminated(
...
-                # Generate another goal
-                goal = self.generate_target_goal()
-                # Add noise to goal position
-                self.goal = self.add_xy_position_noise(goal)
-                # Update the position of the target site for visualization
-                self.update_target_site_pos()
-
             return False
 
+    def update_goal(self, achieved_goal: np.ndarray, desired_goal: np.ndarray) -> None:
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/maze.py:370]: Remove this argument `achieved_goal`,  `desired_goal`
Can be read from member variables

File: gymnasium_robotics/envs/maze/maze.py [hunk 5]
```diff
@@ -363,19 +365,22 @@ def compute_terminated(
...
+        """Update goal position if continuing task and within goal radius."""
+        if (
+            self.continuing_task
+            and bool(np.linalg.norm(achieved_goal - desired_goal) <= 0.45)
+            and len(self.maze.unique_goal_locations) > 1
+        ):
+            # Generate another goal
+            goal = self.generate_target_goal()
+            # Add noise to goal position
+            self.goal = self.add_xy_position_noise(goal)
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/maze.py:380]: Do we need to test if the object I near the goal (0.45)

File: gymnasium_robotics/envs/maze/ant_maze.py [hunk 6]
```diff
@@ -271,21 +272,27 @@ def reset(self, *, seed: Optional[int] = None, **kwargs):
...
         terminated = self.compute_terminated(obs["achieved_goal"], self.goal, info)
         truncated = self.compute_truncated(obs["achieved_goal"], self.goal, info)
 
-        reward = self.compute_reward(obs["achieved_goal"], self.goal, info)
-
         if self.render_mode == "human":
             self.render()
 
+        # Update the goal position if necessary
+        self.update_goal(obs["achieved_goal"], self.goal)
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/ant_maze.py:294]: Should be after `truncated `, before `info`

File: gymnasium_robotics/envs/maze/maze.py [hunk 7]
```diff
@@ -262,7 +262,9 @@ def generate_reset_pos(self) -> np.ndarray:
 
         # While reset position is close to goal position
         reset_pos = self.goal.copy()
-        while np.linalg.norm(reset_pos - self.goal) <= 0.5:
+        while (
+            np.linalg.norm(reset_pos - self.goal) <= 0.5 * self.maze.maze_size_scaling
```
Comment (Kallinteris-Andreas) [inline gymnasium_robotics/envs/maze/maze.py:266]: Shouldn't it be `self.maze._maze_size_scaling`?

**Comment by alexdavey** [comment]
> @alexdavey. We should be keeping previous versions of the environment such as in Gymnasium. You can create different file versions for each environment version `antmaze_v3.py` and `antmaze_v4.py`. The updates in `maze.py` don't have any breaking changes so it doesn't require a version bump.

@rodrigodelazcano. I would have thought that the`maze_size_scaling` bug fix in`maze.py` is a breaking change, since it will change the reset behaviour of the existing AntMaze version?

File: gymnasium_robotics/envs/maze/maze.py (hunk 5 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/maze.py:380]: I agree. @alexdavey can you add a while loop to keep generating a `goal` until the distance to `achieved_goal` is greater than the `0.45` threshold?

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 2 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/ant_maze.py:277]: `info` should be generated for reset as well. Although, this should always be `False` it's good to compute it just to double check. We should include this in the tests.

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 3 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/ant_maze.py:285]: I think the order doesn't matter. Correct me if I'm missing anything @Kallinteris-Andreas

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 2 above)
Comment (alexdavey) [reply gymnasium_robotics/envs/maze/ant_maze.py:277]: Good idea, will add a test.

File: gymnasium_robotics/envs/maze/maze.py (hunk 4 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/maze.py:370]: These variables need to be passed as arguments since a `MazeEnv` object doesn't contain these attributes. They are present in the environment child class which contains the point or ant agents.

File: gymnasium_robotics/envs/maze/maze.py (hunk 7 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/maze.py:266]: This is okey. The attribute is wrapped with `@property` https://github.com/Farama-Foundation/Gymnasium-Robotics/blob/5a13eaae5e56b48c0c61833b7365fe114c249dcb/gymnasium_robotics/envs/maze/maze.py#L71

**Comment by rodrigodelazcano** [comment]
> > @alexdavey. We should be keeping previous versions of the environment such as in Gymnasium. You can create different file versions for each environment version `antmaze_v3.py` and `antmaze_v4.py`. The updates in `maze.py` don't have any breaking changes so it doesn't require a version bump.
> 
> @rodrigodelazcano. I would have thought that the`maze_size_scaling` bug fix in`maze.py` is a breaking change, since it will change the reset behaviour of the existing AntMaze version?

My take on this is that the `maze_size_scaling` is `1` so it won't affect the registered environments

> > @alexdavey. We should be keeping previous versions of the environment such as in Gymnasium. You can create different file versions for each environment version `antmaze_v3.py` and `antmaze_v4.py`. The updates in `maze.py` don't have any breaking changes so it doesn't require a version bump.
> 
> @rodrigodelazcano. I would have thought that the`maze_size_scaling` bug fix in`maze.py` is a breaking change, since it will change the reset behaviour of the existing AntMaze version?
> 
> **Edit:** I have listed what seem like the most sensible version changes to me, below. Let me know if you are happy with these and I will update the PR accordingly:
> 
>     * `maze.py`: The original pre-PR file.
> 
>     * `ant_maze.py`: The original pre-PR file.
> 
>     * `maze_v4.py`: Includes refactor of `compute_terminated()` and the fix for the `maze_size_scaling` bug.
> 
>     * `ant_maze_v4.py`: Uses `maze_v4.py` and fixes sparse reward bug.
> 
>     * `point_maze.py`: Uses `maze_v4.py` but no other changes. No version bump as there should be no behaviour change.

Sounds good. My take on `maze.py` is that `maze_size_scaling` has a value of `1` so it wouldn't affect the registered envs unless you are modifying this argument. However, I agree  lets bump to be sure.

File: tests/envs/maze/test_ant_maze.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_reset():
+    """Check that AntMaze does not reset into a success state."""
+    env = gym.make("AntMaze_UMaze-v4", continuing_task=True)
+
+    for _ in range(1000):
+        obs, _ = env.reset()
+        dist = np.linalg.norm(obs["achieved_goal"] - obs["desired_goal"])
+        assert dist > 0.45, f"dist={dist} < 0.45"
```
Comment (rodrigodelazcano) [inline tests/envs/maze/test_ant_maze.py:12]: @alexdavey can you also assert that the `info['success']` is `False`?

File: tests/envs/maze/test_point_maze.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_reset():
+    """Check that PointMaze does not reset into a success state."""
+    env = gym.make("PointMaze_UMaze-v3", continuing_task=True)
+
+    for _ in range(1000):
+        obs, _ = env.reset()
+        dist = np.linalg.norm(obs["achieved_goal"] - obs["desired_goal"])
+        assert dist > 0.45, f"dist={dist} < 0.45"
```
Comment (rodrigodelazcano) [inline tests/envs/maze/test_point_maze.py:12]: assert for `info['success']`

**Review by rodrigodelazcano — COMMENTED** [review]
I agree with the version changes. I also added some minor comments on @Kallinteris-Andreas review and other updates for the PR., Thank you @alexdavey

**Comment by alexdavey** [comment]
> Sounds good. My take on `maze.py` is that `maze_size_scaling` has a value of `1` so it wouldn't affect the registered envs unless you are modifying this argument. However, I agree lets bump to be sure.

In AntMaze, the MazeEnv parent is initalised with `maze_size_scaling = 4`, so I believe that AntMaze would change behaviour with the new `maze.py` changes.

Thanks to both of you for the reviews.

**Comment by rodrigodelazcano** [comment]
> > Sounds good. My take on `maze.py` is that `maze_size_scaling` has a value of `1` so it wouldn't affect the registered envs unless you are modifying this argument. However, I agree lets bump to be sure.
> 
> In AntMaze, the MazeEnv parent is initalised with `maze_size_scaling = 4`, so I believe that AntMaze would change behaviour with the new `maze.py` changes, but do let me know if I am mistaken.
> 
> Thanks to both of you for the reviews.

You are totally right! Then I take my words back, bumping makes more sense

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 3 above)
Comment (Kallinteris-Andreas) [reply gymnasium_robotics/envs/maze/ant_maze.py:285]: It does not affect the behavior, but makes it more readable

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 6 above)
Comment (alexdavey) [reply gymnasium_robotics/envs/maze/ant_maze.py:294]: I would think that moving `update_goal()` before `info` would mean that `info['success'] = false` always?

Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/ant_maze.py:294]: @alexdavey is right here, we should keep it as it is

File: gymnasium_robotics/envs/maze/ant_maze.py (hunk 3 above)
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/ant_maze.py:285]: Sure. @alexdavey if you can update it as per @Kallinteris-Andreas request that'll be great :)

**Review by rodrigodelazcano — APPROVED** [review]

File: gymnasium_robotics/envs/maze/maze.py [hunk 10]
```diff
@@ -40,6 +40,9 @@ class Maze:
     The Maze class also presents a method to convert from cell indices to `(x,y)` coordinates in the MuJoCo simulation:
     - :meth:`cell_rowcol_to_xy` - Convert from `(i,j)` to `(x,y)`
 
+    ### Version History
```
[the remark this answers is not in the document]
Comment (alexdavey) [reply gymnasium_robotics/envs/maze/maze.py:43]: I will remove the ones I added to `maze.py`. I think `ant_maze.py` has no changes currently from the existing version -- did you want me to remove them anyway?

[the remark this answers is not in the document]
Comment (rodrigodelazcano) [reply gymnasium_robotics/envs/maze/maze.py:43]: As a standard we only keep docstrings in the latest file version. That'll be `ant_maze_v4.py` and `maze_v4.py`. So yes remove them for both `maze,.py` and `ant_maze.py`

[the remark this answers is not in the document]
Comment (alexdavey) [reply gymnasium_robotics/envs/maze/maze.py:43]: Makes sense, will do now

**Comment by rodrigodelazcano** [comment]
Thanks a lot @alexdavey . I'll merge this after the CI checks

**Comment by alexdavey** [comment]
I just removed some final docstrings in `maze.py` that were missed initially. I think all issues have been resolved as well. Thanks.
